File: maestro/leindex/rust/src/cfg.rs
//! CFG Analyzer (Layer 3)
//!
//! Control flow complexity analysis - calculates cyclomatic complexity,
//! identifies decision points, loops, and exception handling.

use serde::{Deserialize, Serialize};
use std::collections::{HashMap, HashSet};

/// Types of nodes in the control flow graph
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum CFGNodeType {
    Entry,
    Exit,
    BasicBlock,
    Condition,      // if/elif
    Loop,           // for/while
    Try,
    Except,
    Finally,
    With,
    Match,          // match statement (Python 3.10+)
}

/// A node in the control flow graph
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CFGNode {
    pub id: String,
    pub node_type: CFGNodeType,
    pub line: usize,
    pub condition: Option<String>,
    pub statements: Vec<String>,
    pub successors: HashSet<String>,
    pub predecessors: HashSet<String>,
}

/// Complexity metrics for a function
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ComplexityMetrics {
    pub cyclomatic_complexity: usize,
    pub decision_points: usize,
    pub loop_count: usize,
    pub branch_count: usize,
    pub try_count: usize,
    pub except_count: usize,
    pub max_nesting_depth: usize,
    pub cognitive_complexity: usize,
}

impl ComplexityMetrics {
    pub fn new() -> Self {
        Self {
            cyclomatic_complexity: 1, // Base complexity is 1
            decision_points: 0,
            loop_count: 0,
            branch_count: 0,
            try_count: 0,
            except_count: 0,
            max_nesting_depth: 0,
            cognitive_complexity: 0,
        }
    }

    /// Get complexity rating as string
    pub fn complexity_score(&self) -> &'static str {
        match self.cyclomatic_complexity {
            0..=5 => "low",
            6..=10 => "moderate",
            11..=20 => "high",
            _ => "very_high",
        }
    }
}

impl Default for ComplexityMetrics {
    fn default() -> Self {
        Self::new()
    }
}

/// Control flow graph for a function
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ControlFlowGraph {
    pub function_name: String,
    pub file_path: String,
    pub start_line: usize,
    pub end_line: usize,
    pub nodes: HashMap<String, CFGNode>,
    pub entry_node: Option<String>,
    pub exit_nodes: HashSet<String>,
    pub metrics: ComplexityMetrics,
}

impl ControlFlowGraph {
    pub fn new(function_name: &str, file_path: &str, start_line: usize) -> Self {
        Self {
            function_name: function_name.to_string(),
            file_path: file_path.to_string(),
            start_line,
            end_line: start_line,
            nodes: HashMap::new(),
            entry_node: None,
            exit_nodes: HashSet::new(),
            metrics: ComplexityMetrics::new(),
        }
    }
}

/// CFG analysis result for a file
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CFGAnalysisResult {
    pub file_path: String,
    pub function_metrics: HashMap<String, ComplexityMetrics>,
    pub total_complexity: usize,
    pub average_complexity: f64,
    pub high_complexity_functions: Vec<String>,
}

/// CFG Analyzer
pub struct CFGAnalyzer {
    _node_counter: usize,
}

impl CFGAnalyzer {
    pub fn new() -> Self {
        Self { _node_counter: 0 }
    }

    /// Analyze a Python source file for control flow complexity
    pub fn analyze(&mut self, source: &str, file_path: &str) -> CFGAnalysisResult {
        let mut result = CFGAnalysisResult {
            file_path: file_path.to_string(),
            function_metrics: HashMap::new(),
            total_complexity: 0,
            average_complexity: 0.0,
            high_complexity_functions: Vec::new(),
        };

        let lines: Vec<&str> = source.lines().collect();

        // Find all function definitions and analyze each
        let mut function_starts: Vec<(String, usize, usize)> = Vec::new(); // (name, start_line, indent)

        for (line_num, raw_line) in lines.iter().enumerate() {
            let line = raw_line.trim();
            let indent = raw_line.len() - raw_line.trim_start().len();

            if line.starts_with("def ") || line.starts_with("async def ") {
                let is_async = line.starts_with("async def ");
                let rest = if is_async { &line[10..] } else { &line[4..] };
                if let Some(paren_pos) = rest.find('(') {
                    let func_name = rest[..paren_pos].trim().to_string();
                    function_starts.push((func_name, line_num + 1, indent));
                }
            }
        }

        // Analyze each function
        for (func_name, start_line, func_indent) in function_starts {
            // Find the function's end line
            let mut end_line = start_line;
            for (line_num, raw_line) in lines.iter().enumerate().skip(start_line) {
                let line = raw_line.trim();
                let indent = raw_line.len() - raw_line.trim_start().len();

                if line.is_empty() || line.starts_with('#') {
                    continue;
                }

                // Function ends when we see a line at same or lower indent
                if indent <= func_indent && line_num > start_line - 1 {
                    break;
                }
                end_line = line_num + 1;
            }

            // Extract function body and analyze
            let metrics = self.analyze_function_body(&lines, start_line, end_line, func_indent);
            result.function_metrics.insert(func_name.clone(), metrics.clone());

            if metrics.cyclomatic_complexity > 10 {
                result.high_complexity_functions.push(func_name);
            }
        }

        // Calculate totals
        result.total_complexity = result
            .function_metrics
            .values()
            .map(|m| m.cyclomatic_complexity)
            .sum();

        if !result.function_metrics.is_empty() {
            result.average_complexity =
                result.total_complexity as f64 / result.function_metrics.len() as f64;
        }

        result
    }

    /// Analyze a function body for complexity metrics
    fn analyze_function_body(
        &mut self,
        lines: &[&str],
        start_line: usize,
        end_line: usize,
        _base_indent: usize,
    ) -> ComplexityMetrics {
        let mut metrics = ComplexityMetrics::new();
        let mut nesting_depth = 0;
        let mut max_nesting = 0;
        let mut indent_stack: Vec<usize> = Vec::new();

        for line_idx in start_line..=end_line.min(lines.len()) {
            if line_idx == 0 || line_idx > lines.len() {
                continue;
            }
            let raw_line = lines[line_idx - 1];
            let line = raw_line.trim();

            if line.is_empty() || line.starts_with('#') {
                continue;
            }

            let indent = raw_line.len() - raw_line.trim_start().len();

            // Track nesting depth
            while !indent_stack.is_empty() && indent <= *indent_stack.last().unwrap() {
                indent_stack.pop();
                if nesting_depth > 0 {
                    nesting_depth -= 1;
                }
            }

            // Check for control flow statements
            if line.starts_with("if ") || line.starts_with("elif ") {
                metrics.decision_points += 1;
                metrics.branch_count += 1;
                metrics.cognitive_complexity += 1 + nesting_depth;
                indent_stack.push(indent);
                nesting_depth += 1;
            } else if line.starts_with("else:") {
                metrics.branch_count += 1;
                // else doesn't add to cyclomatic complexity but adds to cognitive
                metrics.cognitive_complexity += nesting_depth;
            } else if line.starts_with("for ") || line.starts_with("while ") {
                metrics.loop_count += 1;
                metrics.decision_points += 1;
                metrics.cognitive_complexity += 1 + nesting_depth;
                indent_stack.push(indent);
                nesting_depth += 1;
            } else if line.starts_with("try:") {
                metrics.try_count += 1;
                indent_stack.push(indent);
                nesting_depth += 1;
            } else if line.starts_with("except") {
                metrics.except_count += 1;
                metrics.decision_points += 1;
                metrics.cognitive_complexity += nesting_depth;
            } else if line.starts_with("finally:") {
                metrics.cognitive_complexity += nesting_depth;
            } else if line.starts_with("with ") {
                metrics.cognitive_complexity += 1;
                indent_stack.push(indent);
                nesting_depth += 1;
            } else if line.starts_with("match ") {
                metrics.decision_points += 1;
                metrics.cognitive_complexity += 1 + nesting_depth;
                indent_stack.push(indent);
                nesting_depth += 1;
            } else if line.starts_with("case ") {
                metrics.branch_count += 1;
                metrics.cognitive_complexity += nesting_depth;
            }

            // Check for logical operators (add to complexity)
            if line.contains(" and ") {
                let and_count = line.matches(" and ").count();
                metrics.decision_points += and_count;
                metrics.cognitive_complexity += and_count;
            }
            if line.contains(" or ") {
                let or_count = line.matches(" or ").count();
                metrics.decision_points += or_count;
                metrics.cognitive_complexity += or_count;
            }

            // Check for ternary expressions
            if line.contains(" if ") && line.contains(" else ") && !line.starts_with("if ") {
                metrics.decision_points += 1;
                metrics.cognitive_complexity += 1;
            }

            // Check for comprehensions with conditions
            if (line.contains("[") || line.contains("{"))
                && line.contains(" for ")
                && line.contains(" if ")
            {
                metrics.decision_points += 1;
                metrics.cognitive_complexity += 1;
            }

            max_nesting = max_nesting.max(nesting_depth);
        }

        metrics.max_nesting_depth = max_nesting;

        // Calculate cyclomatic complexity: E - N + 2P
        // Simplified: 1 + number of decision points
        metrics.cyclomatic_complexity = 1 + metrics.decision_points;

        metrics
    }

    /// Convert analysis result to LLM-friendly string
    pub fn to_llm_string(&self, result: &CFGAnalysisResult) -> String {
        let mut lines = Vec::new();

        lines.push(format!(
            "# CFG Analysis: {} ({} functions)",
            result.file_path,
            result.function_metrics.len()
        ));

        if !result.function_metrics.is_empty() {
            lines.push(format!(
                "Total Complexity: {}, Average: {:.1}",
                result.total_complexity, result.average_complexity
            ));
        }

        // High complexity functions first
        if !result.high_complexity_functions.is_empty() {
            lines.push(String::new());
            lines.push("## High Complexity Functions (>10)".to_string());
            for func_name in &result.high_complexity_functions {
                if let Some(metrics) = result.function_metrics.get(func_name) {
                    lines.push(format!(
                        "  {} - CC:{} ({}), loops:{}, depth:{}",
                        func_name,
                        metrics.cyclomatic_complexity,
                        metrics.complexity_score(),
                        metrics.loop_count,
                        metrics.max_nesting_depth
                    ));
                }
            }
        }

        // All functions summary
        lines.push(String::new());
        lines.push("## Function Metrics".to_string());

        // Sort by complexity (descending)
        let mut sorted_funcs: Vec<_> = result.function_metrics.iter().collect();
        sorted_funcs.sort_by(|a, b| b.1.cyclomatic_complexity.cmp(&a.1.cyclomatic_complexity));

        for (func_name, metrics) in sorted_funcs.iter().take(20) {
            let complexity_indicator = match metrics.complexity_score() {
                "low" => "",
                "moderate" => "*",
                "high" => "**",
                "very_high" => "***",
                _ => "",
            };

            lines.push(format!(
                "  {}{} CC:{} loops:{} branches:{} depth:{}",
                func_name,
                complexity_indicator,
                metrics.cyclomatic_complexity,
                metrics.loop_count,
                metrics.branch_count,
                metrics.max_nesting_depth
            ));
        }

        if result.function_metrics.len() > 20 {
            lines.push(format!(
                "  ... and {} more functions",
                result.function_metrics.len() - 20
            ));
        }

        lines.join("\n")
    }
}

impl Default for CFGAnalyzer {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_simple_function_complexity() {
        let source = r#"
def simple():
    return 42
"#;
        let mut analyzer = CFGAnalyzer::new();
        let result = analyzer.analyze(source, "test.py");

        assert_eq!(result.function_metrics.len(), 1);
        let metrics = result.function_metrics.get("simple").unwrap();
        assert_eq!(metrics.cyclomatic_complexity, 1);
    }

    #[test]
    fn test_if_else_complexity() {
        let source = r#"
def branching(x):
    if x > 0:
        return 1
    elif x < 0:
        return -1
    else:
        return 0
"#;
        let mut analyzer = CFGAnalyzer::new();
        let result = analyzer.analyze(source, "test.py");

        let metrics = result.function_metrics.get("branching").unwrap();
        assert!(metrics.cyclomatic_complexity >= 3);
        assert!(metrics.branch_count >= 2);
    }

    #[test]
    fn test_loop_complexity() {
        let source = r#"
def looping(items):
    result = []
    for item in items:
        if item > 0:
            result.append(item)
    return result
"#;
        let mut analyzer = CFGAnalyzer::new();
        let result = analyzer.analyze(source, "test.py");

        let metrics = result.function_metrics.get("looping").unwrap();
        assert!(metrics.loop_count >= 1);
        assert!(metrics.max_nesting_depth >= 1);
    }

    #[test]
    fn test_nested_complexity() {
        let source = r#"
def nested(matrix):
    for row in matrix:
        for cell in row:
            if cell > 0:
                print(cell)
"#;
        let mut analyzer = CFGAnalyzer::new();
        let result = analyzer.analyze(source, "test.py");

        let metrics = result.function_metrics.get("nested").unwrap();
        assert!(metrics.max_nesting_depth >= 2);
    }
}

File: maestro/archive/leindex-python/analyzers/cfg.py
import ast
from dataclasses import dataclass, field, asdict
from typing import Optional, List, Dict, Set, Any, Tuple
from enum import Enum
from .base import BaseAnalyzer

class NodeType(str, Enum):
    """Types of nodes in the control flow graph"""
    ENTRY = "entry"
    EXIT = "exit"
    BASIC_BLOCK = "basic_block"
    CONDITION = "condition"
    LOOP = "loop"
    TRY = "try"
    EXCEPT = "except"
    FINALLY = "finally"

@dataclass
class CFGNode:
    """A node in the control flow graph"""
    id: str
    type: NodeType
    line: int
    condition: Optional[str] = None
    statements: List[str] = field(default_factory=list)
    successors: Set[str] = field(default_factory=set)
    predecessors: Set[str] = field(default_factory=set)

@dataclass
class ComplexityMetrics:
    """Complexity metrics for a function"""
    cyclomatic_complexity: int = 1
    decision_points: int = 0
    loop_count: int = 0
    branch_count: int = 0
    try_count: int = 0
    except_count: int = 0
    max_nesting_depth: int = 0

    def complexity_score(self) -> str:
        """Get complexity rating"""
        if self.cyclomatic_complexity <= 5:
            return "low"
        elif self.cyclomatic_complexity <= 10:
            return "moderate"
        elif self.cyclomatic_complexity <= 20:
            return "high"
        else:
            return "very_high"

@dataclass
class ControlFlowGraph:
    """Complete control flow graph for a function"""
    function_name: str
    file_path: str
    start_line: int
    end_line: int
    nodes: Dict[str, CFGNode] = field(default_factory=dict)
    entry_node: Optional[str] = None
    exit_nodes: Set[str] = field(default_factory=set)
    metrics: ComplexityMetrics = field(default_factory=ComplexityMetrics)

class CFGAnalyzer(BaseAnalyzer):
    """
    Control Flow Graph Analyzer.
    Layer 3 analysis: complexity metrics, control flow patterns.
    """

    def __init__(self) -> None:
        self._node_counter = 0

    def analyze(self, code: str, file_path: str) -> Dict[str, Any]:
        """
        Analyze code to calculate CFG metrics for all functions.
        """
        try:
            tree = ast.parse(code)
        except SyntaxError:
             return {
                "path": file_path,
                "error": "SyntaxError"
            }

        function_metrics = {}

        for node in ast.walk(tree):
            if isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef)):
                cfg = self._build_cfg(node, file_path)
                if cfg:
                    function_metrics[node.name] = asdict(cfg.metrics)
                    function_metrics[node.name]['score'] = cfg.metrics.complexity_score()

        return {
            "path": file_path,
            "functions": function_metrics
        }

    def _build_cfg(self, function_node: ast.AST, file_path: str) -> Optional[ControlFlowGraph]:
        self._node_counter = 0
        cfg = ControlFlowGraph(
            function_name=function_node.name,
            file_path=file_path,
            start_line=function_node.lineno,
            end_line=function_node.end_lineno or function_node.lineno,
        )

        # Build the CFG
        entry_id = self._new_node_id()
        cfg.entry_node = entry_id
        cfg.nodes[entry_id] = CFGNode(id=entry_id, type=NodeType.ENTRY, line=function_node.lineno)

        # Process function body
        exit_ids = self._process_block(function_node.body, cfg, entry_id)

        # Create common exit node
        exit_id = self._new_node_id()
        cfg.exit_nodes.add(exit_id)
        cfg.nodes[exit_id] = CFGNode(id=exit_id, type=NodeType.EXIT, line=function_node.end_lineno or function_node.lineno)

        # Connect all exit paths to common exit
        for ex in exit_ids:
            if ex in cfg.nodes:
                cfg.nodes[ex].successors.add(exit_id)
                cfg.nodes[exit_id].predecessors.add(ex)

        # Calculate metrics
        cfg.metrics = self._calculate_metrics(cfg)

        return cfg

    def _new_node_id(self) -> str:
        self._node_counter += 1
        return f"n{self._node_counter}"

    def _process_block(
        self,
        statements: List[ast.stmt],
        cfg: ControlFlowGraph,
        entry_id: str,
    ) -> Set[str]:
        if not statements:
            return {entry_id}

        current_exits = {entry_id}

        for stmt in statements:
            new_exits = set()
            for exit_id in current_exits:
                if isinstance(stmt, ast.If):
                    new_exits.update(self._process_if(stmt, cfg, exit_id))
                elif isinstance(stmt, (ast.For, ast.While)):
                    new_exits.update(self._process_loop(stmt, cfg, exit_id))
                elif isinstance(stmt, ast.Try):
                    new_exits.update(self._process_try(stmt, cfg, exit_id))
                elif isinstance(stmt, ast.Return):
                    new_exits.update(self._process_return(stmt, cfg, exit_id))
                else:
                    node = cfg.nodes.get(exit_id)
                    if node and node.type == NodeType.BASIC_BLOCK:
                        node.statements.append(ast.unparse(stmt)[:100])
                        new_exits.add(exit_id)
                    else:
                        new_id = self._new_node_id()
                        cfg.nodes[new_id] = CFGNode(
                            id=new_id,
                            type=NodeType.BASIC_BLOCK,
                            line=stmt.lineno,
                            statements=[ast.unparse(stmt)[:100]]
                        )
                        cfg.nodes[exit_id].successors.add(new_id)
                        cfg.nodes[new_id].predecessors.add(exit_id)
                        new_exits.add(new_id)

            current_exits = new_exits

        return current_exits

    def _process_if(self, if_node: ast.If, cfg: ControlFlowGraph, entry_id: str) -> Set[str]:
        cond_id = self._new_node_id()
        cfg.nodes[cond_id] = CFGNode(
            id=cond_id,
            type=NodeType.CONDITION,
            line=if_node.lineno,
            condition=ast.unparse(if_node.test)[:100],
        )

        cfg.nodes[entry_id].successors.add(cond_id)
        cfg.nodes[cond_id].predecessors.add(entry_id)

        then_exits = self._process_block(if_node.body, cfg, cond_id)

        if if_node.orelse:
            else_exits = self._process_block(if_node.orelse, cfg, cond_id)
        else:
            else_exits = {cond_id}

        return then_exits | else_exits

    def _process_loop(self, loop_node: ast.For | ast.While, cfg: ControlFlowGraph, entry_id: str) -> Set[str]:
        loop_id = self._new_node_id()
        condition = ast.unparse(loop_node.iter if isinstance(loop_node, ast.For) else loop_node.test)[:100]
        cfg.nodes[loop_id] = CFGNode(
            id=loop_id,
            type=NodeType.LOOP,
            line=loop_node.lineno,
            condition=condition,
        )

        cfg.nodes[entry_id].successors.add(loop_id)
        cfg.nodes[loop_id].predecessors.add(entry_id)

        body_exits = self._process_block(loop_node.body, cfg, loop_id)

        for exit_id in body_exits:
            if exit_id in cfg.nodes:
                cfg.nodes[exit_id].successors.add(loop_id)
                cfg.nodes[loop_id].predecessors.add(exit_id)

        if loop_node.orelse:
            else_exits = self._process_block(loop_node.orelse, cfg, loop_id)
        else:
            else_exits = {loop_id}

        return else_exits

    def _process_try(self, try_node: ast.Try, cfg: ControlFlowGraph, entry_id: str) -> Set[str]:
        try_id = self._new_node_id()
        cfg.nodes[try_id] = CFGNode(
            id=try_id,
            type=NodeType.TRY,
            line=try_node.lineno,
        )

        cfg.nodes[entry_id].successors.add(try_id)
        cfg.nodes[try_id].predecessors.add(entry_id)

        try_exits = self._process_block(try_node.body, cfg, try_id)

        all_exits = set(try_exits)
        for handler in try_node.handlers:
            except_id = self._new_node_id()
            exc_type = ast.unparse(handler.type) if handler.type else "Exception"
            cfg.nodes[except_id] = CFGNode(
                id=except_id,
                type=NodeType.EXCEPT,
                line=handler.lineno,
                condition=f"except {exc_type}",
            )
            cfg.nodes[try_id].successors.add(except_id)
            cfg.nodes[except_id].predecessors.add(try_id)

            except_exits = self._process_block(handler.body, cfg, except_id)
            all_exits.update(except_exits)

        if try_node.finalbody:
            finally_id = self._new_node_id()
            cfg.nodes[finally_id] = CFGNode(
                id=finally_id,
                type=NodeType.FINALLY,
                line=try_node.finalbody[0].lineno,
            )

            for exit_id in all_exits.copy():
                if exit_id in cfg.nodes:
                    cfg.nodes[exit_id].successors.add(finally_id)
                    cfg.nodes[finally_id].predecessors.add(exit_id)
                    all_exits.remove(exit_id)

            finally_exits = self._process_block(try_node.finalbody, cfg, finally_id)
            all_exits.update(finally_exits)

        return all_exits

    def _process_return(self, return_node: ast.Return, cfg: ControlFlowGraph, entry_id: str) -> Set[str]:
        ret_id = self._new_node_id()
        cfg.nodes[ret_id] = CFGNode(
            id=ret_id,
            type=NodeType.EXIT,
            line=return_node.lineno,
            statements=["return"],
        )

        cfg.nodes[entry_id].successors.add(ret_id)
        cfg.nodes[ret_id].predecessors.add(entry_id)
        cfg.exit_nodes.add(ret_id)

        return {ret_id}

    def _calculate_metrics(self, cfg: ControlFlowGraph) -> ComplexityMetrics:
        metrics = ComplexityMetrics()

        for node in cfg.nodes.values():
            if node.type == NodeType.CONDITION:
                metrics.decision_points += 1
                metrics.branch_count += 2
            elif node.type == NodeType.LOOP:
                metrics.loop_count += 1
                metrics.decision_points += 1
            elif node.type == NodeType.TRY:
                metrics.try_count += 1
            elif node.type == NodeType.EXCEPT:
                metrics.except_count += 1

        metrics.cyclomatic_complexity = 1 + metrics.decision_points + metrics.except_count
        metrics.max_nesting_depth = self._calculate_nesting_depth(cfg)

        return metrics

    def _calculate_nesting_depth(self, cfg: ControlFlowGraph) -> int:
        depth = 0
        # Use a max recursion depth safeguard or iterative approach if needed,
        # but for simple CFGs DFS is fine. We need to avoid cycles.
        visited = set()

        def dfs(node_id: str, current_depth: int):
            nonlocal depth
            if node_id in visited:
                return
            visited.add(node_id)

            depth = max(depth, current_depth)

            node = cfg.nodes.get(node_id)
            if not node:
                return

            for succ in node.successors:
                extra = 1 if node.type in (NodeType.CONDITION, NodeType.LOOP, NodeType.TRY) else 0
                dfs(succ, current_depth + extra)

            visited.remove(node_id)

        if cfg.entry_node:
            dfs(cfg.entry_node, 0)

        return depth

    def to_llm_string(self, analysis_result: Dict[str, Any]) -> str:
        lines = [f"File: {analysis_result['path']}"]

        for func_name, metrics in analysis_result.get("functions", {}).items():
            lines.append(f"\nFunction: {func_name}")
            lines.append(f"  Complexity: {metrics['cyclomatic_complexity']} ({metrics['score']})")
            if metrics['decision_points']:
                lines.append(f"  Decision Points: {metrics['decision_points']}")
            if metrics['loop_count']:
                lines.append(f"  Loops: {metrics['loop_count']}")

        return "\n".join(lines)
